joern-server/src/main/scala/io/shiftleft/joern/server/JoernController.scala

```diff
@@ -0,0 +1,158 @@
+    extends ScalatraServlet
+    with NativeJsonSupport
+    with SwaggerSupport
+    with FutureSupport {
+
+  val logger = LoggerFactory.getLogger(getClass)
+  protected implicit val jsonFormats: Formats = DefaultFormats
+  protected val applicationDescription = "Joern-Server REST API"
+  protected implicit def executor: ExecutionContext = system.dispatcher
+
+  private var cpg: Option[Cpg] = None
+  private var queryResult: Option[String] = None
+
+  before() {
+    contentType = formats("json")
+  }
+
+  private val createBuilder =
+    (apiOperation[Unit]("create")
+      summary "Create a code property graph"
+      tags ""
+      parameter queryParam[List[String]]("filenames").description("File/Directory names"))
+
+  post("/create", operation(createBuilder)) {
+    handleOrReportAsInvalid[CreateCpgRequest] { v =>
+      val filenames = v.filenames
+      if (filenames.isEmpty) {
+        halt(400, "`filenames` not given or invalid")
+      }
+      if (filenames.count(File(_).exists) != filenames.size) {
+        halt(400, "Not all specified files exist")
+      }
+      createCpg(filenames)
+      response.setHeader("Location", s"/status")
+      Accepted()
+    }
+
+    def createCpg(filenames: List[String]): AsyncResult =
+      new AsyncResult {
+        val is = Future {
+          val cpgFilename = "/tmp/cpg.bin.zip"
+          logger.info(s"Attempting to create CPG for: ${filenames.mkString(",")}")
+          JoernParse.parse(filenames.toArray, cpgFilename)
+          cpg = Some(CpgLoader.load(cpgFilename))
+          logger.info("CPG is ready")
+        }
+      }
+  }
+
+  private val statusBuilder = (apiOperation[String]("status")
+    summary "Status of the CPG"
+    tags "")
+
+  get("/status", operation(statusBuilder)) {
+    Ok(StatusResponse(isCpgLoaded = cpg.isDefined))
+  }
+
+  private val queryBuilder = (apiOperation[String]("query")
+    summary "query"
+    tags "")
+
+  post("/query", operation(queryBuilder)) {
+    handleOrReportAsInvalid[QueryRequest] { v =>
+      if (!cpg.isDefined) {
+        BadRequest("CPG is not loaded")
+      } else {
+        val query = v.query
+        logger.info(s"received query: $query")
+        runQuery(query)
+        Accepted()
+      }
+    }
+
+    def runQuery(query: String): AsyncResult =
+      new AsyncResult {
+        val is = Future {
+          Try({
+            logger.info("Running query")
+            import javax.script.ScriptEngineManager
+            val e = new ScriptEngineManager().getEngineByName("scala")
+            e.put("aCpg", cpg.get)
+            e.eval(s"""
+                import io.shiftleft.codepropertygraph.Cpg
+              | val cpg = aCpg.asInstanceOf[io.shiftleft.codepropertygraph.Cpg]
+              | $query
+            """.stripMargin).toString
+          }) match {
+            case Success(v)         => queryResult = Some(v)
+            case Failure(exception) => println(exception)
```

Comment:
this looks a little fishy: the exception message won't make it to the client I guess, and it'll just respond with 'ok' rather than a result?

Reply:
This is run in the Future after the client has already received the OK, so, I can't really pass it back to the client via HTTP though. What I can do though is place it in queryResult as well. That way, the client sees it when picking up the result.